driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java

```diff
@@ -157,13 +164,23 @@ public void abortTransaction(final SingleResultCallback<Void> callback) {
                     new SingleResultCallback<Void>() {
                         @Override
                         public void onResult(final Void result, final Throwable t) {
+                            if (t != null && t instanceof MongoException) {
+                                unpinServerAddressOnError((MongoException) t);
+                            }
                             cleanupTransaction(TransactionState.ABORTED);
                             callback.onResult(null, null);
                         }
                     });
         }
     }
 
+    private void unpinServerAddressOnError(final MongoException e) {
+        if (getPinnedServerAddress() != null
```

Comment:
Nit: It's harmless to unpin an already-unpinned server, so you can simplify the conditional by removing the null check.

Reply:
I have removed the null check.